### PR TITLE
norman: EMA decay warmup schedule on PR #99 base

### PR DESCRIPTION
**Pull request by morganmcg1**
# norman — EMA decay warmup schedule on PR #99 6L/256d base

## Hypothesis

Current PR #99 uses fixed EMA decay = 0.9995 throughout training. With lr=5e-4 driving rapid early progress, a constant decay 0.9995 means the EMA tracks ~2000-step running average — too slow for the first epoch (you're still learning fast), too fast at the end (you want to lock in the converged weights). The `--ema-decay-start` and `--ema-decay-end` flags exist in `train.py` but have not been tested on PR #99. A schedule that starts loose (0.99 — fast tracking) and ramps to tight (0.9999 — heavy averaging at convergence) is standard practice for EMA in modern training (e.g. EDM, Karras et al.) and tends to help especially when epoch budget is tight.

## Baseline to beat (PR #99, fern, W&B run `3hljb0mg`)

| Metric | yi best | AB-UPT |
|---|---:|---:|
| `abupt_axis_mean_rel_l2_pct` | **10.69** | — |

## Experiment plan — 3-arm EMA schedule sweep

Use `--wandb-group norman-ema-warmup-r5`:

| Arm | start | end |
|---|---:|---:|
| A | 0.99 | 0.9995 |
| B | 0.99 | 0.9999 |
| C | 0.999 | 0.9999 |

```bash
cd target/
python train.py \
  --volume-loss-weight 2.0 --batch-size 8 --validation-every 1 \
  --lr 5e-4 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 6 --model-hidden-dim 256 --model-heads 4 --model-slices 128 \
  --ema-decay-start <START> --ema-decay-end <END> \
  --clip-grad-norm 1.0 \
  --wallshear-y-weight 2.0 --wallshear-z-weight 2.0 \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms \
  --wandb-group norman-ema-warmup-r5
```

(Note: omit `--ema-decay 0.9995` — schedule flags should override the static value. If they conflict, please report it.)

## Diagnostics

- Final val_primary metrics for all 3 arms
- Compare EMA model val metrics vs raw model val metrics — if the gap is small, EMA is too slow; if EMA is much better, EMA is helping
- Loss slope at cutoff for each arm

## Reporting

Comment with a 3-arm table; flag the best arm; note whether EMA scheduling matters more than the final decay value.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT norman:

## Bug fix: EMA cosine schedule was effectively flat

While running the 3-arm EMA schedule sweep, I found a bug in the existing `--ema-decay-start`/`--ema-decay-end` cosine schedule that invalidates the experiment as originally designed. Reporting and fixing here so the corrected sweep is meaningful.

### The bug

The schedule denominator is pegged to the nominal `--epochs` flag (default 50), but our wall-clock timeout (`SENPAI_TIMEOUT_MINUTES=360`) only allows ~4 epochs at this config (one epoch ≈ 85 min). So the cosine progress only reaches ~8% of its arc, and the decay barely budges from `--ema-decay-start`.

`train.py:1815-1820` (before fix):
```python
total_estimated_steps = max(1, max_epochs * max(len(train_loader), 1))   # 50 * 10883 = 544k
progress = min(global_step / max(total_estimated_steps, 1), 1.0)
cos_val = (1.0 - math.cos(math.pi * progress)) / 2.0
ema_decay_now = config.ema_decay_start + cos_val * (config.ema_decay_end - config.ema_decay_start)
```

Empirical confirmation from the broken first launch (group `norman-ema-warmup-r5`):

| Arm | step | logged `train/ema_decay` |
|---|---:|---:|
| A (0.99→0.9995) | 5243 | 0.99000218 |
| B (0.99→0.9999) | 5211 | 0.99000224 |
| C (0.999→0.9999) | 5212 | 0.99900020 |

After ~½ epoch the decay had moved ~0.000002 of its intended ~0.01 swing. Projected at the end of training (~43k steps at 4 epochs), arms would land at A=0.99015, B=0.99020, C=0.99901 — i.e. essentially constant at their start values, not actual schedules. Arms A and B would be functionally identical.

### Fix

Added `--ema-decay-schedule-epochs N` (default `0` = preserve existing behavior using `max_epochs`). When > 0, the cosine ramp denominator uses `N * len(train_loader)` instead of `max_epochs * len(train_loader)`, so the schedule completes in `N` epochs of training:

```python
ema_schedule_epochs = (
    config.ema_decay_schedule_epochs if config.ema_decay_schedule_epochs > 0 else max_epochs
)
ema_schedule_total_steps = max(1, ema_schedule_epochs * max(len(train_loader), 1))
...
progress = min(global_step / ema_schedule_total_steps, 1.0)
```

This is additive and backward-compatible — existing runs that don't set the flag are unaffected.

### Re-launch

I killed the 3 broken runs and re-launched as group `norman-ema-warmup-r6` with `--ema-decay-schedule-epochs 4` so the cosine completes around the wall-clock cutoff. Same 3 arms (A: 0.99→0.9995, B: 0.99→0.9999, C: 0.999→0.9999). New W&B run IDs:

- Arm A: `yd78evfy`
- Arm B: `1a9j05ob`
- Arm C: `hq2h3jz7`

Will post results when training completes.

**Comment by morganmcg1** [comment]
STUDENT norman:

## Arm A diverged in epoch 1 — re-launching

Arm A (`--ema-decay-start 0.99 --ema-decay-end 0.9995`, run `yd78evfy`) diverged during epoch 1 with classic gradient explosion:

| step | pre_clip grad norm |
|---:|---:|
| 4899 | 140 |
| 9299 | 168k |
| 9399 | 1.6M |
| 9499 | **1.3T** |
| 9599 | nan (4.7M nonfinite params) |

epoch 1 finished with `train_loss=nan`, `val_*=nan`. Arm B (max grad norm = 67) and arm C (max grad norm = 24) showed no instability — same hyperparameters except EMA decay end. Since EMA does not affect training gradients, this is stochastic divergence (random init / batch order luck), not a hyperparameter issue.

Killed PID 173523 and re-launched arm A on GPU 0:
- Arm A retry: run `4sl61huk` (started 2026-04-30 23:26)

Arms B (`1a9j05ob`) and C (`hq2h3jz7`) continue normally. Will report all 3 arms when they complete.

**Comment by morganmcg1** [comment]
STUDENT norman:

## Arm A diverged again — relaunching

Arm A retry (`4sl61huk`) diverged at step ~9300 with the same gradual-grad-explosion pattern as the original arm A:

| step | train_loss | pre_clip_grad_norm |
|---:|---:|---:|
| 7800 | 0.04 | 1.0 |
| 8400 | 0.23 | 10.6 |
| 8800 | 1.12 | 11.9 |
| 9000 | 2.71 | 44.2 |
| 9100 | 2.29 | 130 |
| 9200 | 2.22 | 414 |
| 9300 | NaN | NaN |

EMA cannot affect training gradients (it's `@torch.no_grad()` after `optimizer.step()`), so this isn't an EMA-schedule issue — it's training instability around mid-epoch-1 with lr=5e-4 + clip=1.0. The grad clipping is being applied, but the loss drifts upward over hundreds of steps before final explosion. Arms B (start=0.99 same as A) and C (start=0.999) have shown no instability, so it's stochastic.

I killed PID 190419 and relaunched arm A on GPU 0 as `norman/ema-warmup-A-099-9995-r6-retry2`. If this third attempt also diverges I will report results for arms B and C only with arm A flagged as unstable.

- Arm A retry-2: launching now (group `norman-ema-warmup-r6`)
- Arms B (`1a9j05ob`) and C (`hq2h3jz7`) continue normally — both at epoch 3 with healthy grad norms (~1).

Will report all results when training completes.

**Comment by morganmcg1** [comment]
ADVISOR: Acknowledged on the third arm A retry. The pattern is clear — this is NOT about EMA config (your analysis confirming EMA is @no_grad is correct), it's stochastic instability from a rare bad batch triggering gradient explosion at lr=5e-4/clip=1.0. Arms B/C are healthy. Here's what to do: let the third retry run — if it stabilizes, great; if it NaN's again before epoch 1 val, kill it and report B/C only. Don't waste more retries on A if it fails a third time. The arms B (ema-start=0.999) and C (ema-start=0.9995) finishing healthy with 3+ epochs is the core result regardless of A. When all arms complete, post final test_primary/* and mark ready for review.

**Comment by morganmcg1** [comment]
STUDENT norman:

## Results

3-arm EMA decay-schedule sweep on the PR #99 6L/256d base. Bug fix for the broken cosine schedule (using `--ema-decay-schedule-epochs 4`) is included in this PR; details in the earlier comment.

Group: `norman-ema-warmup-r6`.

### Headline (best EMA-model checkpoint, all metrics from the same checkpoint reload)

| Arm | EMA start → end | Best epoch | val_primary `abupt_axis_mean_rel_l2_pct` | test_primary `abupt_axis_mean_rel_l2_pct` |
|---|---|:---:|---:|---:|
| **PR#99 baseline** (constant 0.9995) | — | 4 | **10.69** | **11.73** |
| A | 0.99 → 0.9995 | — | NaN | NaN |
| B | 0.99 → 0.9999 | 2 | 11.54 | 12.59 |
| **C** | **0.999 → 0.9999** | **4** | **10.84** | **11.93** |

- Arm A diverged 3 times (3rd retry NaN'd at step ~6.4k). Reported and stopped per advisor.
- Arm B trained healthily through epoch 3 (val 11.54) then NaN'd mid-epoch-4 around step ~31.4k; checkpoint from epoch 2 was used for test eval (post-hoc, run id `1a9j05ob`).
- Arm C completed 4 full epochs (272 min) with no instability.

### Test-primary breakdown (best checkpoint)

| Metric | AB-UPT ref | PR#99 baseline | Arm B (ep2) | Arm C (ep4) |
|---|---:|---:|---:|---:|
| `abupt_axis_mean_rel_l2_pct` | — | 11.73 | 12.59 | **11.93** |
| `surface_pressure_rel_l2_pct` | 3.82 | 6.64 | 7.51 | **6.79** |
| `wall_shear_rel_l2_pct` | 7.29 | 11.48 | 12.73 | **11.82** |
| `volume_pressure_rel_l2_pct` | 6.08 | 14.42 | 13.91 | **14.12** |
| `wall_shear_x_rel_l2_pct` | 5.35 | — | 11.23 | 10.33 |
| `wall_shear_y_rel_l2_pct` | 3.65 | — | 14.82 | 13.95 |
| `wall_shear_z_rel_l2_pct` | 3.63 | — | 15.49 | 14.47 |
| `surface_pressure_mae` | — | 0.01878 | 0.02124 | 0.01916 |
| `wall_shear_mae` | — | 0.10388 | 0.11664 | 0.10706 |
| `volume_pressure_mae` | — | 30.487 | 30.317 | 30.496 |

(Bold = best across arms.)

### Loss slope at cutoff (`train/slope/loss/per_1k_steps`)

- Arm B: −0.135 (still going down at the moment of NaN explosion)
- Arm C: −0.134 (healthy convergence)

### What happened

**The hypothesis did not hold within the 4-epoch wall-clock budget.** The best schedule arm (C, 0.999 → 0.9999) lost to PR #99's constant 0.9995 by **+0.20 pct on test_primary** and +0.15 pct on val_primary. The intuition that loose-then-tight EMA helps is reasonable, but with only 4 epochs of effective training the constant 0.9995 averages enough updates throughout — the schedule just under-averages early (start=0.999 means ~1k-step running avg) and barely over-averages at the very end. Net effect: small loss vs. a well-tuned constant.

**Both arms with `start=0.99` (A and B) suffered gradient-explosion divergence**, while the arm with `start=0.999` (C) was stable. EMA cannot affect gradients (the `ema.update(model)` call is `@torch.no_grad()` and runs *after* `optimizer.step()`), so this is genuinely stochastic — same lr, same clip_grad_norm=1.0 between A/B and C, just bad init/batch luck on the lower-decay arms. Three NaN's out of three start=0.99 attempts (counting both arm A retries plus arm B at epoch 4) is a striking coincidence but mechanistically not EMA-driven.

**The EMA model still tracks the best-checkpoint val gap reasonably well** — arm C's saved (EMA) checkpoint reloaded to val 10.84/test 11.93, and arm B's saved checkpoint reloaded to val 11.54/test 12.59 (separate post-hoc eval), confirming the EMA shadow weights remain useful even though arm B's raw training later NaN'd.

### Suggested follow-ups

- **Tighter schedule range, longer ramp**: try start=0.9995, end=0.9998 over `--ema-decay-schedule-epochs 6` so the sweep covers a wider window of the decay range without the risky low-decay early phase. The current 0.999→0.9999 covers <1 order of magnitude in averaging horizon.
- **Bigger epoch budget**: schedule probably matters more when training is long enough that the late-stage convergence regime exists. At 4 epochs the model is still in the rapid-improvement phase; the late-stage benefit of high decay never gets exercised.
- **Don't go below 0.999 with this lr**: arm A and arm B's start=0.99 didn't help and the runs were unstable. Floor at 0.999 for this 5e-4 / clip=1.0 setup.
- **Bug fix to merge regardless**: the `--ema-decay-schedule-epochs` flag (this PR) is independently useful — the prior schedule was effectively constant when the wall-clock budget allowed only ~8% of the nominal `--epochs` arc. Future EMA-schedule experiments need it to work.

### Train commands

Arm A (failed; for reference):
```bash
python train.py \
  --volume-loss-weight 2.0 --batch-size 8 --validation-every 1 \
  --lr 5e-4 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 6 --model-hidden-dim 256 --model-heads 4 --model-slices 128 \
  --ema-decay-start 0.99 --ema-decay-end 0.9995 \
  --ema-decay-schedule-epochs 4 \
  --clip-grad-norm 1.0 \
  --wallshear-y-weight 2.0 --wallshear-z-weight 2.0 \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms \
  --agent norman --wandb-group norman-ema-warmup-r6 \
  --wandb-name norman/ema-warmup-A-099-9995-r6
```

Arm B and C are identical except for `--ema-decay-start` / `--ema-decay-end` (0.99/0.9999 and 0.999/0.9999 respectively) and `--wandb-name`.

Post-hoc test eval for arm B (epoch-2 checkpoint) was run via a small script that imported `train.py`'s `evaluate_split`, loaded the saved EMA-baked checkpoint, and resumed into the original W&B run id.

### Run IDs and runtime

| Arm | W&B run | Status | Wall-clock |
|---|---|---|---:|
| A (3 attempts) | `yd78evfy`, `4sl61huk`, `giy4aagj` | all NaN'd in epoch 1 | <50 min each |
| B | `1a9j05ob` | NaN'd at step ~31.4k (epoch 4) — epoch-2 checkpoint used | 4h 23m before kill |
| C | `hq2h3jz7` | completed 4 epochs cleanly | 272 min |

Peak GPU memory: **75.5 GB** on a single H100 (logged in `[*GB]` epoch lines).

**Comment by morganmcg1** [comment]
ADVISOR: Thank you for this thorough diagnostic work, norman. I reviewed the W&B results directly.

**Verified W&B Results:**
- Arm B (EMA 0.999→0.9990, run `1a9j05ob`): test_primary/abupt_axis_mean = **12.593%** 
- Arm C (EMA 0.999→0.9999, run `hq2h3jz7`): test_primary/abupt_axis_mean = **11.932%** (best arm)
- Baseline (PR #99): **11.73%**

**Decision: CLOSE — does not beat baseline.**

Arm C regresses by +0.20 pct vs baseline (11.932 vs 11.73). The EMA decay schedule hypothesis does not hold within the 4-epoch wall-clock budget: your bug report is correct that the cosine schedule was nearly flat (denominator pegged to `--epochs=50` while we only complete ~4 epochs), but even with a corrected schedule the sweep shows no benefit — and mildly hurts.

**However: the `--ema-decay-schedule-epochs` bug fix is independently valuable.** The bug means every EMA-sweep experiment on bengio has been running with a flat schedule rather than the intended cosine. I will cherry-pick this fix onto the bengio branch.

**Suggested next direction:** The wsy/wsz channels are universally 3-4× above AB-UPT targets (wsy ~13-14% vs target 3.65, wsz ~14-15% vs target 3.63). Consider experimenting with per-channel loss weighting — specifically upweighting wall_shear_y and wall_shear_z relative to the other outputs — as these are the channels where the model has the most headroom. A sweep of `surface_loss_weight` by decomposed head or a learned task weighting (uncertainty-weighted multi-task loss) could be productive.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #128 because EMA decay schedule does not beat baseline within 4-epoch wall-clock budget. Arm C (best): test_abupt=11.932% vs baseline 11.73% — regression. Bug fix for --ema-decay-schedule-epochs will be cherry-picked to bengio branch separately..
